diff -Naur nethack-3.4.3/src/files.c nethack-3.4.3-bones/src/files.c
--- nethack-3.4.3/src/files.c	2009-03-22 11:54:23.000000000 -0400
+++ nethack-3.4.3-bones/src/files.c	2009-03-22 11:56:17.000000000 -0400
@@ -57,7 +57,7 @@
 #endif
 
 #if !defined(MFLOPPY) && !defined(VMS) && !defined(WIN32)
-char bones[] = "bonesnn.xxx";
+char bones[] = "bones/bonesnn.xxx";
 char lock[PL_NSIZ+14] = "1lock"; /* long enough for uid+name+.99 */
 #else
 # if defined(MFLOPPY)
@@ -614,8 +614,13 @@
 	s_level *sptr;
 	char *dptr;
 
+#ifdef UNIX
+	Sprintf(file, "bones/bon%c%s", dungeons[lev->dnum].boneid,
+			In_quest(lev) ? urole.filecode : "0");
+#else
 	Sprintf(file, "bon%c%s", dungeons[lev->dnum].boneid,
 			In_quest(lev) ? urole.filecode : "0");
+#endif
 	dptr = eos(file);
 	if ((sptr = Is_special(lev)) != 0)
 	    Sprintf(dptr, ".%c", sptr->boneid);
diff -Naur nethack-3.4.3/sys/unix/Makefile.top nethack-3.4.3-bones/sys/unix/Makefile.top
--- nethack-3.4.3/sys/unix/Makefile.top	2003-12-07 18:39:13.000000000 -0500
+++ nethack-3.4.3-bones/sys/unix/Makefile.top	2009-03-22 11:57:54.000000000 -0400
@@ -234,12 +234,12 @@
 # set up the directories
 # not all mkdirs have -p; those that don't will create a -p directory
 	-mkdir -p $(SHELLDIR)
-	-rm -rf $(GAMEDIR) $(VARDIR)
-	-mkdir -p $(GAMEDIR) $(VARDIR) $(VARDIR)/save
+	#-rm -rf $(GAMEDIR) $(VARDIR)
+	-mkdir -p $(GAMEDIR) $(VARDIR) $(VARDIR)/save $(VARDIR)/bones
 	-rmdir ./-p
-	-$(CHOWN) $(GAMEUID) $(GAMEDIR) $(VARDIR) $(VARDIR)/save
-	$(CHGRP) $(GAMEGRP) $(GAMEDIR) $(VARDIR) $(VARDIR)/save
-	chmod $(DIRPERM) $(GAMEDIR) $(VARDIR) $(VARDIR)/save
+	-$(CHOWN) $(GAMEUID) $(GAMEDIR) $(VARDIR) $(VARDIR)/save $(VARDIR)/bones
+	$(CHGRP) $(GAMEGRP) $(GAMEDIR) $(VARDIR) $(VARDIR)/save $(VARDIR)/bones
+	chmod $(DIRPERM) $(GAMEDIR) $(VARDIR) $(VARDIR)/save $(VARDIR)/bones
 # set up the game files
 	( $(MAKE) dofiles )
 # set up some additional files
